**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №10

Набор данных: auto.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

In [ ]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df = pd.read_csv('auto.csv')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [6]:
# применить метод head
print(df.head(21))

                               Name  year  SellingPrice  kmdriven     fuel  \
0                     Maruti 800 AC  2007       60000.0   70000.0   Petrol   
1          Maruti Wagon R LXI Minor  2007      135000.0   50000.0   Petrol   
2              Hyundai Verna 1.6 SX  2012      600000.0  100000.0   Diesel   
3            Datsun RediGO T Option  2017      250000.0   46000.0   Petrol   
4             Honda Amaze VX i-DTEC  2014      450000.0  141000.0   Diesel   
5              Maruti Alto LX BSIII  2007      140000.0  125000.0   Petrol   
6         Hyundai Xcent 1.2 Kappa S  2016      550000.0   25000.0   Petrol   
7         Tata Indigo Grand Petroll  2014      240000.0   60000.0   Petrol   
8          Hyundai Creta 1.6 VTVT S  2015      850000.0   25000.0  Petroll   
9          Maruti Celerio Green VXI  2017      365000.0   78000.0      CNG   
10          Chevrolet Sail 1.2 Base  2015      260000.0   35000.0   Petrol   
11         Tata Indigo Grand Petrol  2014      250000.0  100000.

2.2 Оценка данных с помощью метода info.

In [4]:
# выполнит метод info
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4344 entries, 0 to 4343
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          4344 non-null   object 
 1   year          4344 non-null   int64  
 2   SellingPrice  4342 non-null   float64
 3   kmdriven      4343 non-null   float64
 4   fuel          4343 non-null   object 
 5   seller_Type   4344 non-null   object 
 6   transmission  4344 non-null   object 
 7   owner         4344 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 186.7+ KB
None


2.3 Оценка данных с помощью метода describe.

In [7]:
# оцените числовые столбцы с помощью describe
print(df.describe())

              year  SellingPrice       kmdriven
count  4344.000000  4.342000e+03    4343.000000
mean   2013.092311  5.039987e+05   66193.063320
std       4.214543  5.784465e+05   46636.648764
min    1992.000000  2.000000e+04       1.000000
25%    2011.000000  2.100000e+05   35000.000000
50%    2014.000000  3.500000e+05   60000.000000
75%    2016.000000  6.000000e+05   90000.000000
max    2020.000000  8.900000e+06  806599.000000



---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


### Общая информация о датасете

1. В датасете содержится **4344** строки и **8** столбцов
2. Типы данных:
    + **5 столбцов** имеют тип данных *'object'* (это текстовые данные, например название автомобиля, тип топлива)
    + **2 столбца** имеют тип *'float64'* (это числовые значения с плавающей точкой (`SellingPrice`, `kmdriven`))
    + **1 столбец** имеет тип *'int64'* (это целочисленный год выпуска (`year`))
3. Почти все значения заполнены, но:
   + В столбце `SellingPrice` отсутствуют **2 значения**
   + В столбце `kmdriven` отсутствует **1 значение**
   + В столбце `fuel` отсутствует **1 значение**

Это говорит о наличии **единичных пропусков**, которые нужно будет обработать на следующем этапе.

### Статистический обзор числовых столбцов

1. **Год выпуска (`year`)**:
      - Диапазон: от **1992** до **2020** года.
      - Средний год: **2013**, медианный — **2014**.
      - Это говорит о том, что большинство автомобилей — относительно новые, с преобладанием моделей после 2010 года.

2. **Цена продажи (`SellingPrice`)**:
      - Диапазон: от **20,000** до **8,900,000** (вероятно, в рублях).
      - Средняя цена: около **503,998**.
      - Распределение цен сильно варьируется, что может указывать на наличие как бюджетных, так и премиальных автомобилей.

3. **Пробег (`kmdriven`)**:
      - Диапазон: от **1 км** до **806,599 км**.
      - Средний пробег: около **66,193 км**, медианный — **60,000 км**.
      - Большинство автомобилей имеют пробег в пределах **35,000–90,000 км**, что считается нормальным для подержанных машин.

Эти данные помогут в дальнейшем выявить выбросы, сегментировать автомобили по возрасту и пробегу, а также построить модели оценки стоимости.

 2.4 Оценка названий столбцов

In [8]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
print(df.columns)

Index(['Name', 'year', 'SellingPrice', 'kmdriven', 'fuel', 'seller_Type',
       'transmission', 'owner'],
      dtype='object')


### Анализ названий столбцов

В целом, большинство названий понятны, но выявлены следующие проблемы:
- Столбец `kmdriven` содержит опечатку. Корректное название — `km_driven`.
- Столбец `seller_Type` использует смешанный стиль написания. Рекомендуется привести к единому стилю — `snake_case`, например `seller_type`.
- Название `Name` можно уточнить, чтобы оно отражало суть — например, `car_name` или `model`.
- Название `SellingPrice` выполнено в стиле `camelCase`, нужно исправить в `snake_case`.

In [11]:
# переименование при необходимости
df.rename(columns={
    'kmdriven': 'km_driven',
    'seller_Type': 'seller_type',
    'Name': 'car_name',
    'SellingPrice': 'selling_price'
}, inplace=True)

In [12]:
# Проверим результат выполнения команды
print(df.columns)

Index(['car_name', 'year', 'selling_price', 'km_driven', 'fuel', 'seller_type',
       'transmission', 'owner'],
      dtype='object')


### 3. Проверка пропусков

In [ ]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [ ]:
# удалите дубликаты, если они есть

#### Проверка неявных дубликатов

In [ ]:
# удалите дубликаты, если они есть

---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [ ]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Формулировка задания`*

In [ ]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*

In [ ]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

In [ ]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

In [ ]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

Вывод должен описывать результаты исследования: какой набор данных вы анализировали (предметная область), какие методы предобработки применили, какие выводы из группировок и сводных таблиц можно сделать (они могут повторять выводы выше) и т.п. Объём вывода, как правило не более 0.5-0.75 страницы А4, всё таки вывод должен быть конкретным (без теории и “воды”).


### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***